### ISSUE TITLE
WebGL

### ISSUE THREAD

**Issue by sh7dm** [denoland/deno#1629]
Idea: create a GPGPU library or even standart functions for Deno.

**Comment by kyranet** [comment]
I believe this is out of the scope for Deno... we do not need to implement every library in core, but you can make one and PR a link in [registry](https://github.com/denoland/registry)

**Comment by ry** [comment]
Not out of scope. I want to support WebGL in core. But i won't be ready to start this for many months.

**Comment by sh7dm** [comment]
@kyranet It needs native bindings, you can't use OpenCL/CUDA in JS :laughing:

**Comment by kyranet** [comment]
I assume this would slowly lead to a native implementation of `canvas` in Deno, right? I find this a very complicated task, but also one of the most interesting things 😄 

And sorry about my previous message, I wasn't very sure of what @sh7dm meant with GPGPU 👍

**Comment by ry** [comment]
The plan is to use Angle: https://cs.chromium.org/chromium/src/third_party/angle/BUILD.gn

following the example of https://github.com/stackgl/headless-gl

**Comment by sh7dm** [comment]
@ry it renders its output to file?

**Comment by KeKsBoTer** [comment]
@sh7dm rendering to an image is one option. But you can also create a window with Angle, as seen in this example: https://cs.chromium.org/chromium/src/third_party/angle/samples/WindowTest/WindowTest.cpp?g=0

**Comment by quininer** [comment]
Although still not a standard, [webgpu](https://github.com/gfx-rs/wgpu) is also an interesting choice.

**Comment by sh7dm** [comment]
@quininer interesting

**Comment by Brooooooklyn** [comment]
I will have a try to implement a minimal demo with https://github.com/rust-skia/rust-skia and https://github.com/pcwalton/surfman when I had time.

**Comment by hayd** [comment]
aside: rust-skia looks like a good candidate to use https://github.com/denoland/cargo_gn

**Comment by Uzume** [comment]
Instead of supporting WebGL directly in core (or at least in addition to), it might make more sense to consider [WebGPU](https://webgpu.io/). I realize it is early in the process for such things but we are already on the bleeding edge for several other things like WASM ESM, etc.

**Comment by chances** [comment]
Considering WebGPU, I'm working on bindings to [wgpu-rs](https://github.com/gfx-rs/wgpu-rs) as [a Deno plugin](https://github.com/chances/deno-wgpu/tree/deno).

**Comment by mko-io** [comment]
@maierfelix has been made a [webgpu-node](https://github.com/maierfelix/webgpu) implementation based on Google's Dawn working with Nodejs, and even made[ ray-tracing extension](https://github.com/maierfelix/dawn-ray-tracing) works!!. better integrate his excellent work

**Comment by KeKsBoTer** [comment]
@mko-io I would prefer the [webgpu rust implementation ](https://github.com/gfx-rs/wgpu) since deno is also written in rust. 
One could also argue that Google's Dawn will work better with v8.
IMO [wgpu-rs](https://github.com/gfx-rs/wgpu) is still the better choice for deno.

**Comment by crowlKats** [comment]
I think [Glow](https://github.com/grovesNL/glow) would be the right way to go for WebGL support

**Comment by DjDeveloperr** [comment]
Are there still plans to implement WebGL considering WebGPU has landed already?

**Comment by crowlKats** [comment]
WebGL can technically be polyfilled using webgpu, but that's quite the task, but i think that would be the best solution, as we wouldn't need to depend on a second graphics api

**Comment by lucacasonato** [comment]
I'm going to close this. We have WebGPU now, which is going to be capable of everything WebGL is. There is no point in implementing both APIs (like fetch and XXLHttpRequest).

**Comment by Uzume** [comment]
I am assuming this is closed due to the recent merge of  #7977.

**Comment by chirsz-ever** [comment]
> I am assuming this is closed due to #7863 and the recent merge of #7977. Now we just need a WebGL polyfill (implemented using the WebGPU API). It might be possible to do this in TypeScript/JavaScript but a Rust implementation could be interesting too.

Do you have any good idea about "WebGL polyfill over WebGPU"? I am interested in such a thing. WebGL need a canvas for create context, and the corresponding thing in deno is [`Deno.UnsafeWindowSurface `](https://deno.land/api@v1.42.4?s=Deno.UnsafeWindowSurface)? We can use the polyfill like this:

```ts
// modified from https://deno.com/blog/v1.40#webgpu-windowing--bring-your-own-window
import {
  EventType,
  WindowBuilder,
} from "https://deno.land/x/sdl2@0.8.0/mod.ts";

const win = new WindowBuilder("Hello, World!", 800, 600).build();

/* Returns a Deno.UnsafeWindowSurface */
const surface = win.windowSurface();
/* Returns a WebGL WebGLRenderingContext */
const gl = surface.getContext("webgl");
gl.Clear(gl.COLOR_BUFFER_BIT);
```

Is this good?
